clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -3104,20 +3104,58 @@ static void handleWorkGroupSize(Sema &S, Decl *D, const ParsedAttr &AL) {
       return;
   }
 
-  if (WorkGroupAttr *ExistingAttr = D->getAttr<WorkGroupAttr>()) {
-    ASTContext &Ctx = S.getASTContext();
+  ASTContext &Ctx = S.getASTContext();
+
+  if (!XDimExpr->isValueDependent() || !YDimExpr->isValueDependent() ||
+      !ZDimExpr->isValueDependent()) {
     Optional<llvm::APSInt> XDimVal = XDimExpr->getIntegerConstantExpr(Ctx);
     Optional<llvm::APSInt> YDimVal = YDimExpr->getIntegerConstantExpr(Ctx);
     Optional<llvm::APSInt> ZDimVal = ZDimExpr->getIntegerConstantExpr(Ctx);
-    Optional<llvm::APSInt> ExistingXDimVal = ExistingAttr->getXDimVal(Ctx);
-    Optional<llvm::APSInt> ExistingYDimVal = ExistingAttr->getYDimVal(Ctx);
-    Optional<llvm::APSInt> ExistingZDimVal = ExistingAttr->getZDimVal(Ctx);
 
-    // Compare attribute arguments value and warn for a mismatch.
-    if (ExistingXDimVal != XDimVal || ExistingYDimVal != YDimVal ||
-        ExistingZDimVal != ZDimVal) {
-      S.Diag(AL.getLoc(), diag::warn_duplicate_attribute) << AL;
-      S.Diag(ExistingAttr->getLocation(), diag::note_conflicting_attribute);
+    if (AL.getKind() == ParsedAttr::AT_ReqdWorkGroupSize) {
+      if (const auto *A = D->getAttr<SYCLIntelNumSimdWorkItemsAttr>()) {
+        int64_t NumSimdWorkItems =
+            A->getValue()->getIntegerConstantExpr(Ctx)->getSExtValue();
+
+        assert(NumSimdWorkItems && XDimVal &&
+               "Argument should be an integer constant expression");
+        assert(NumSimdWorkItems && YDimVal &&
+               "Argument should be an integer constant expression");
+        assert(NumSimdWorkItems && ZDimVal &&
+               "Argument should be an integer constant expression");
+
+        if (NumSimdWorkItems == 0)
+          return;
```

Comment:
This means we're silently dropping the reqd work group size attribute with no diagnostic. I think the correct behavior here is, when processing the `SYCLIntelNumSimdWorkItemsAttr`, check to see if the value is 0, and if it is, issue a diagnostic and don't attach the attribute AST node to the declaration. In that case, this check can become an assertion.

Reply:
Done